fix(material/a11y_assessments): add unique page title

essjay05 opened the pull request:
Adds usage of SystemChrome.setAPplicationSwitcherDescription to update the page title for the AutoComplete demo page.

[Before screenshot: example AutoComplete](https://screenshot.googleplex.com/8DpESj4QrQA2YrW)
[After screenshot: example AutoComplete](https://screenshot.googleplex.com/6AdfSA8a8VwUaXM)

Fixes b/338056490

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

essjay05 committed:
fix(material/auto_complete): add unique page title

Adds usage of SystemChrome.setAPplicationSwitcherDescription
to update the page title for the AutoComplete demo page.

Fixes b/338056490

essjay05 committed:
fix(flutter/a11y_assessments): attempting to provide dynamic page titles per use case

Creates a setWebTitle in a utils.dart to be used in all use case pages to update the
page title.

Fixes b/338056490

essjay05 committed:
fix(flutter/a11y_assessments): give each page a unique page title

Updates each use_case to have a unique page title time.

essjay05 committed:
refactor(flutter/a11y_assessments): fixing lint error

Adds new line underneath last snippet in utils.dart.

chunhtai commented on dev/a11y_assessments/lib/main.dart:
```diff
@@ -80,6 +82,8 @@ class HomePageState extends State<HomePage> {
 
   @override
   Widget build(BuildContext context) {
+    setWebTitle('Home Page', Theme.of(context).colorScheme.primary.value);
```
Should we use Title widget instead?

essjay05 replied:
Updating to use Title widget instead.

chunhtai replied:
This is tangential, but should we produce warning if a web app has never set title before? Is there a use case you don't want change title at all?

essjay05 replied:
> This is tangential, but should we produce warning if a web app has never set title before? Is there a use case you don't want change title at all?

To meet GAR standards, each page should have a unique title, which is why I'm attempting to add one for each of the use case pages.

I think potential good warnings to have are if there's no title set, and if each "new page" does not have a title or different title.

chunhtai reviewed (commented):
you can test using
```
tester.binding.defaultBinaryMessenger.setMockMethodCallHandler(SystemChannels.platform, (MethodCall methodCall) async {
      // verify the method call with expected value
    });
```
just search `tester.binding.defaultBinaryMessenger.setMockMethodCallHandler` across code base, you should find a bunch of examples

essjay05 committed:
refactor(flutter/a11y_assessments): ipmlementing title widget as recommended

Updates previous fix to use Title widget instead of SystemChrome.

essjay05 committed:
test(flutter/a11y_assessments): adding unique page title tests for a11y_assessments use cases

Updates previous fix to check for each use case having its own unique page title test.

essjay05 committed:
refactor(flutter/a11y_assessments): updating use_case page title

Updates how to implement each use_case page title.

essjay05 committed:
refactor(flutter/a11y_assessments): refactoring add title and adding test to home page

Updates previous changes to home page where we give the home page a title and also
adds a test to check for an existing title.

essjay05 committed:
rebase(flutter/a11y_assessments): rebase to text_button changes from https://github.com/flutter/flutter/pull/152185

Updates my PR with merged changes from PR 1528185.

essjay05 committed:
fix(flutter/a11y_assessments): added title widget wrapper around new use_cases

Updates new use_case demos a Title widget to apply custom/unique page title.

essjay05 committed:
test(flutter/a11y_assessments): successfully added check title tests for some pages

Created check for unique page title test and created successful tests for home page,
autocomplete use case, and checkbox list title use case. WIP: remaining use_cases.

essjay05 committed:
test(flutter/a11y_assessments): added all tests for use_case page titles

Updates home_page_test file to check each use_case page for its actual page title
(which were manually entered in and intentionally unique).

essjay05 commented:
> you can test using
> 
> ```
> tester.binding.defaultBinaryMessenger.setMockMethodCallHandler(SystemChannels.platform, (MethodCall methodCall) async {
>       // verify the method call with expected value
>     });
> ```
> 
> just search `tester.binding.defaultBinaryMessenger.setMockMethodCallHandler` across code base, you should find a bunch of examples

Updated the home_page_test.dart file by checking each use_case page for its respective title page.

essjay05 committed:
fix(flutter/a11y_assessment): fix linux failing test unused import

Updates previous commit which adds all tests and removes the unused import
from the main.dart file.

essjay05 committed:
fix(flutter/a11y_assessments): fix lint errors

Add recommended header to utils.dart file.

chunhtai commented on dev/a11y_assessments/lib/utils.dart:
```diff
@@ -0,0 +1,7 @@
+
+import 'package:flutter/services.dart';
+
+Color appSeedColor = const Color(0xff6750a4);
```
```suggestion
const Color appSeedColor = Color(0xff6750a4);
```

chunhtai replied:
also why a hardcode value vs using something like Theme.of(context).primaryColor

essjay05 replied:
Updated to use Theme.of(context).primary.

chunhtai replied:
in that case is this still needed?

essjay05 replied:
Removed appSeedColor

chunhtai commented on dev/a11y_assessments/lib/use_cases/text_field.dart:
```diff
@@ -22,40 +22,44 @@ class _MainWidget extends StatelessWidget {
-                hintText: 'Enter your email',
+    return Title(
+      color: appSeedColor,
+      title: 'TextField Demo',
```
should we use UseCase.name directly?

essjay05 replied:
Updated to use UseCase.name as recommended and not hardcode color.

essjay05 committed:
refactor(flutter/a11y_assessments): update fix based on review

Updates previous fix based on suggestions by reviewer. Also
adds title and tests for new components. Updated conflicts.

essjay05 committed:
test(flutter/a11y_assessments): refactor action chip and add missing h1 tests

Updates previous change to refactor action_chip to use pageTitle and
add missing h1 tests.

chunhtai commented on dev/a11y_assessments/lib/use_cases/action_chip.dart:
```diff
@@ -26,27 +26,34 @@ class MainWidget extends StatefulWidget {
-              label: Text('Action'),
-            ),
-          ],
+    return Title(
```
Instead of doing this in individual use case, I think you can either do this in UseCase base class or in main.dart when it creates the routes for each use case, it can also wrap the title automatically there

essjay05 replied:
Updated to create a DynamicTitle widget that wraps the routes and updates the page title based on the use case name value.

essjay05 replied:
Hi @chunhtai oddly enough after updating the DynamicTitle to wrap as you recommended this now breaks the test cases. Please see my latest changes in this commit: ([72c080d](https://github.com/flutter/flutter/pull/152148/commits/72c080dc9e5e3ab53252ffb089b224fcd624c127). 

The log shows that the Check for the main page works, but when it pumps the AutoComplete that the label comes up empty: 
![image](https://github.com/user-attachments/assets/dfcc520a-cbec-49d9-866f-583a864e0fd8)

I'm still tinkering with it to try to see why it's not pulling it but thought I'd commit it so I could reference it.

essjay05 replied:
Hi @chunhtai you can disregard my previous note. I found the fix (had to update the pumpsUseCase in test_utils.dart). The working and latest changes can be found in this commit: [02758a3](https://github.com/flutter/flutter/pull/152148/commits/02758a38e9a63f28d00d1908a0e2d8f2779b40f4).

essjay05 committed:
refactor(flutter/a11y_assessments): create reusable DynamicTitle wrapper

Updates previous fix to create a pageTitle widget wrapper around the routes
and each use_case updates the page title based on the useCase name prop.
Updates test to new page title name.

chunhtai commented on dev/a11y_assessments/lib/use_cases/action_chip.dart:
```diff
@@ -26,12 +26,15 @@ class MainWidget extends StatefulWidget {
 class MainWidgetState extends State<MainWidget> {
   bool favorite = false;
+
+  String pageTitle = getUseCaseName(ActionChipUseCase());
```
is this still used?

chunhtai replied:
same for other use cases

essjay05 replied:
This is being used to pull the Use case name value into the h1 heading in the AppBar on line 37 (and same with all the other use cases).

chunhtai commented on dev/a11y_assessments/lib/main.dart:
```diff
@@ -37,11 +37,46 @@ class App extends StatelessWidget {
+
+  @override
+  Widget build(BuildContext context) {
+    return DynamicTitle(
```
For home page it will use MaterialApp(title directly), I don't think you need to wrap the home page here?

essjay05 replied:
Updated based on your suggestions.

chunhtai commented on dev/a11y_assessments/lib/use_cases/use_cases.dart:
```diff
@@ -28,7 +29,14 @@ import 'text_field_password.dart';
   String get name;
   String get route;
-  Widget build(BuildContext context);
+  Widget get mainWidget;
```
why change this to a widget instead of a builder?

you can have a wrapper like this
```dart
Widget myActualBuild(BuildContext context) {
      return DynamicTitle(
      title: name,
      child: build(context),
    );
}
```

essjay05 committed:
refactor(flutter/a11y_assessments): move DynamicTitle to wrap useCase

Updates previous fix to wrap the DynamicTitle around the useCase as
recommended in PR review. WIP: New changes have caused the tests
checking for the unique page title on each use case page are now
showing empty labels. Committing to reference to reviewer.

essjay05 committed:
tests(flutter/a11y_assessments): updated use_case tests checking for page title

Updated test_utils.dart to mimic changes to main.dart to ensure DynamicTitle
is implemented into test cases. Updated home_page_test.dart file for checking
page title to make sure the correct log index is set to its respective page
title string.

essjay05 committed:
refactor(flutter/a11y_assessments): reformat unaliagned code in test

Updates previous commit to fix formatting to make code more readable.

chunhtai commented on dev/a11y_assessments/lib/common/dynamic_title.dart:
```diff
@@ -0,0 +1,17 @@
+      child: child,
+    );
+  }
+}
```
a new line at the end of the file

chunhtai commented on dev/a11y_assessments/lib/use_cases/use_cases.dart:
```diff
@@ -28,6 +29,15 @@ import 'text_field_password.dart';
 abstract class UseCase {
   String get name;
   String get route;
+  // Widget get mainWidget;
```
remove this

chunhtai commented on dev/a11y_assessments/lib/use_cases/use_cases.dart:
```diff
@@ -28,6 +29,15 @@ import 'text_field_password.dart';
   String get route;
+  // Widget get mainWidget;
+
+  Widget myActualBuild(BuildContext context) {
```
sorry I used a bad name in my previous reply. Maybe call this buildWithTitle.

essjay05 replied:
Thanks for the rec! I've updated to the new recommended descriptive name.

chunhtai reviewed (approved):
LGTM, just some minor comments

essjay05 committed:
refactor(fix/a11y_assessments): update with recommended fixes from PR review

Updates previous fix to make build that wraps use_cases to be more descriptive
and also removes unused code. Fixes lint errors.

essjay05 committed:
fix(flutter/a11y_assessments): fixes rebase error of added title

Updates dialog.dart which rebased to pull previous code wrapping
a Title widget around the Dialog widget. This fixes the failing
test.

essjay05 committed:
fix(flutter/a11y_assessments): fix missing license lines

Updates newly created dynamic_title.dart file to include the
right license header for dart files.
